### PR TITLE
add tableacl proto and refactor go/vt/tableacl

### PR DESCRIPTION
**Pull request by yaoshengzhe**
In short, the tableacl configuration will be represented as a
tableacl.proto. tableacl module holds a global "currentACL" instance
which maps a table group to its ACLs.

1. Introduce tableacl.proto which defines acl per table group.
   A table group either contains only one table or many tables
   that share the same name prefix.
2. Remove "All" and "AllString" funcs from acl.Factory interface.

### CODE REVIEW

File: go/vt/tableacl/tableacl.go [hunk 1]
```diff
@@ -39,72 +68,82 @@ func Init(configFile string) {
...
+// Authorized returns the list of entities who have the specified role on a tablel.
+func Authorized(table string, role Role) acl.ACL {
+	currentACL.RLock()
+	defer currentACL.RUnlock()
+	start := 0
+	end := len(currentACL.entries)
+	for start < end {
+		mid := start + (end-start)/2
+		val := currentACL.entries[mid].tableNameOrPrefix
+		if table == val || (strings.HasSuffix(val, "%") && strings.Contains(table, val[:len(val)-1])) {
```
Comment (alainjobart) [inline go/vt/tableacl/tableacl.go]: should use 'StartsWith' not contain here?

File: proto/tableacl.proto [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+syntax = "proto3";
+
+package tableacl;
+
+// TableGroupSpec defines ACLs for a group of tables.
+message TableGroupSpec {
+  string name = 1;
+  // either tables or a table name prefixes (if it ends in a %)
+  string table_names_or_prefixes = 2;
```
Comment (alainjobart) [inline proto/tableacl.proto]: Should all these fields be repeated?

File: go/vt/tableacl/tableacl.go [hunk 3]
```diff
@@ -39,72 +68,82 @@ func Init(configFile string) {
...
+	config := &tableaclpb.Config{}
+	if err := proto.Unmarshal(data, config); err != nil {
+		return err
 	}
-	tableAcl := make(map[*regexp.Regexp]map[Role]acl.ACL)
-	for tblPattern, accessMap := range contents {
-		re, err := regexp.Compile(tblPattern)
+	var entries aclEntries
+	for _, group := range config.TableGroups {
+		readers, err := newACL(strings.Split(group.Readers, ","))
```
Comment (alainjobart) [inline go/vt/tableacl/tableacl.go]: OK now reading this, let's use repeated fields in the proto, and not split here, you'll get a []string directly.

File: proto/tableacl.proto (hunk 2 above)
Comment (yaoshengzhe) [inline proto/tableacl.proto]: Done. Make all fields in TableGroupSpec (except name field) be repeated.

File: go/vt/tableacl/tableacl.go (hunk 3 above)
Comment (yaoshengzhe) [inline go/vt/tableacl/tableacl.go]: Done.

File: go/vt/tableacl/tableacl.go (hunk 1 above)
Comment (yaoshengzhe) [inline go/vt/tableacl/tableacl.go]: Done. Good catch, use strings.HasPrefix instead.

**Comment by alainjobart** [comment]
LGTM now, but don't forget to add the validation step of the entries, or we're gonna suffer. :)

**Pull request merged**